### ISSUE TITLE
Custom `artifactDirectory` has to exist and is not automatically created (like `__generate__`)

### ISSUE THREAD

**Issue by LukasKalbertodt** [facebook/relay#3782]
When specifying a custom `artifactDirectory` in the relay compiler config, ​the compiler requires this directory to exist. Otherwise you get:

```
[ERROR] Config `/home/.../frontend/relay.config.js` is invalid:
​- The `artifactDirectory` does not exist at `/home/.../frontend/src/query-types`.
```

This is particularly annoying when using git, as you cannot check empty folders into git. The typical "`.gitkeep` file in folder" trick does not work well, because the relay-compiler deletes all files from the artifact directory on each compile, always leading to a dirty working directory.

I think the best is to use the same behavior as for the default `__generated__` folder: create it, if it doesn't exist.

**Comment by alunyov** [comment]
Thanks for report. This is a duplicate of https://github.com/facebook/relay/issues/3727.

We're working on a solution for this.

**Comment by alunyov** [comment]
Adding README/.gitkeep files to `artifactDirectory` should be possible with this https://github.com/facebook/relay/commit/0b5a8b97b4c6bc938998885873bc127964c50785 change. Should be available as `relay-compiler@main` on NPM.

**Comment by LukasKalbertodt** [comment]
Thanks for the quick PR. However, this only enables a workaround, right? This is only to make relay not delete `.gitkeep`. Wouldn't the better solution be to create the artifact directory? I mean, that's the behavior for the default `__generated__`: it is created automatically. I don't see why a custom directory shouldn't.

**Comment by LukasKalbertodt** [comment]
Mh okay, that doesn't really convince me to be honest :/
Lots of other stuff in the config could be misspelled and relay has no way to check that. And further, if the artifact directory is really misspelled, imports in JS code will fail, which will be swiftly noticed by programmers and caught by CI. 

Even with the `.gitkeep` workaround, relative custom artifact directories are very inconvenient to use, as you have to create lots of folders and `.gitkeep` files. So I still think relay-compiler should create those directories.

**Comment by alunyov** [comment]
> Even with the .gitkeep workaround, relative custom artifact directories are very inconvenient to use, as you have to create lots of folders and .gitkeep

Hmm, why do you need to create lots of folders? There is only one folder you need to create if `artifactDirectory` is specified in the config. Otherwise, compiler automatically create artifacts next to the source code, in the `__generated__/MyFile.graphql.js` folder. These `__generated__` do not have to be created by users. That is the main (very small) difference between these two options.

Also, If someone wants to help improve this and implement automatic `artifactDirectory` creation we are happy to review and merge.

**Comment by LukasKalbertodt** [comment]
But if my `artifactDirectory` is relative,  e.g. `"foo"`, I would expect the folder to be located relative to the source file defining the query, no? Like, if not specifying `artifactDirectory`, there are potentially lots of `__generated__`, for each directory that contains source files defining queries. In the project I'm working now:

```
$ tree | grep __generated__
│   ├── __generated__
│   │   ├── __generated__
│   │   ├── __generated__
│   │   │   ├── __generated__
│   │   │   │   │   │   │   ├── __generated__
│   │   │   │   │   │   ├── __generated__
│   │   │   │   │   ├── __generated__
│   │   │   │   ├── __generated__
│   │   │       ├── __generated__
│   │   │   ├── __generated__
```

I want exactly that behavior, but `foo` instead of `__generated__`. That's what I'm talking about. 

---

But frankly, I am not really invested in this issue anymore since we just use the default in our project now. I still think that specifying `artifactDirectory` should change nothing about the semantics, except that the literal folder name `__generated__` is replaced with whatever I specify.

**Comment by alunyov** [comment]
I don't think we have (or had in previous versions) an option to change the default name of the `__generated__` directory.

I'll keep this open, if someone want's to look into this.

**Comment by matthewoates** [comment]
I agree that the compiler should create the directory or at least provide an option to do so. There's a pretty straightforward workaround though:

```js
// package.json
  "scripts": {
    "relay": "mkdir -p __generated__ && relay-compiler --watch",
    "relay-build": "mkdir -p __generated__ && relay-compiler",
    // ...
  },
```

**Comment by sibelius** [comment]
this still a problem
